compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala

```diff
@@ -957,6 +963,64 @@ abstract class PrepJSInterop extends plugins.PluginComponent
       case ctor: MethodSymbol if ctor.isPrimaryConstructor => ctor
     }
 
+  private def needsFixOrCleanup(template: Template): Boolean = {
+    val body = template.body
+    if (body == null) {
+      false
+    } else {
+      body.exists { member =>
+        val sym = member.symbol
+        if (sym != null) sym.hasAnnotation(WasPublicBeforeTyperClass)
+        else false
+      }
+    }
+  }
+
+  private def fixPublicBeforeTyper(template: Template): Template = {
+    val body = template.body
+
+    def needsCleanUp(vddef: ValOrDefDef): Boolean =
+      vddef.symbol.hasAnnotation(WasPublicBeforeTyperClass)
+
+    val accessorsNeedFix = body.collect {
+      case vdef: ValDef if needsCleanUp(vdef) => vdef.name.toString.trim
+    }.toSet
+
+    def needsFix(ddef: DefDef): Boolean = {
+      lazy val isAccessorThatNeedsFix = {
+        val accessorName = ddef.name.toString
+        val fieldName = {
+          if (accessorName.endsWith("_$eq"))
+            accessorName.substring(0, accessorName.length - 4)
+          else
+            accessorName
+        }
+        accessorsNeedFix(fieldName)
+      }
+      val sym = ddef.symbol
+      sym.isPrivate && isJSAny(template.symbol.owner) &&
+      (sym.isAccessor && isAccessorThatNeedsFix) ||
+        sym.hasAnnotation(WasPublicBeforeTyperClass)
+    }
+
+    val newBody = body.map {
+      case ddef: DefDef if needsFix(ddef) =>
+        val sym = ddef.symbol
+        sym.filterAnnotations(!_.matches(WasPublicBeforeTyperClass))
+        sym.resetFlag(Flag.PRIVATE)
+        treeCopy.DefDef(ddef, ddef.mods &~ Flag.PRIVATE, ddef.name,
+            ddef.tparams, ddef.vparamss, ddef.tpt, ddef.rhs)
+
+      case vddef: ValOrDefDef if needsCleanUp(vddef) =>
+        vddef.symbol.filterAnnotations(!_.matches(WasPublicBeforeTyperClass))
+        vddef
+
+      case member =>
+        member
+    }
+
+    treeCopy.Template(template, template.parents, template.self, newBody)
+  }
```

Comment:
What is preventing you to do this when you match a `DefDef` / `ValDef` in the main transformer? It seems you are reimplementing a transform here...